### DISCUSSION TITLE
Letting people ignore compiler flags? Byproduct support?

### DISCUSSION THREAD

**Discussion by 13steinj** [ccache/ccache#1301] [Ideas]
I'm reading the manual and I may be missing it but it seems like this isn't an option (or my language doesn't match my searches).

As far as I know, currently, the entire command line string ends up hashed and used to determine if the cache can be hit.

There's some (newer) compiler flags that don't affect (or at least, realistically shouldn't) affect object code output.

For the sake of immediate example, `-ftime-trace`, which realistically should have equivalent object code output, and produces byproduct output that's purely informational. Similar goes for the externis gcc plugin, as well as `-ftime-report`

I don't know _how_ this would be implemented, since modern argument order is a bit non-deterministic if using a system like cmake (one day you could add an argument and suddenly the argument index changes). Maybe regex-based transform (as well as log the match) of the command line -> command line "that should be hashed"?
Ex one could set a transform of `"-ftime-trace(=\S+)?(\s|$)"` -> `"$2"` to wipe that argument from the command line before hashing.

Of course the before and after transform would have to be logged somewhere, and completely disabled by default (since most options are not purely informational).

---

Maybe I should start another discussion for this bit, but other, non-informational by-products (ex split-dwarf debugging); I wonder if that should somehow be supported for caching.

**Comment by jrosdahl** [comment]
> I'm reading the manual and I may be missing it but it seems like this isn't an option (or my language doesn't match my searches).

There are [`ignore_options`/`CCACHE_IGNOREOPTIONS`](https://ccache.dev/manual/4.8.2.html#config_ignore_options) and [`--ccache-skip`](https://ccache.dev/manual/4.8.2.html#_extra_options) with different effects. I think what you're looking for would be satisfied by making `ignore_options` also skip matching compiler option from special processing (see also https://github.com/ccache/ccache/discussions/1284#discussioncomment-5864486).

> As far as I know, currently, the entire command line string ends up hashed and used to determine if the cache can be hit.

That's partly correct, but some options are not included in the input hash and some other information than the command line is also included. This is described in more details in [_How ccache works_](https://ccache.dev/manual/4.8.2.html#_how_ccache_works).

> Maybe I should start another discussion for this bit, but other, non-informational by-products (ex split-dwarf debugging); I wonder if that should somehow be supported for caching.

Ccache already has support for `-gsplit-dwarf` and caching the resulting file since [version 3.2.3](https://ccache.dev/releasenotes.html#_ccache_3_2_3).

**Reply by jrosdahl** [reply]
> I think what you're looking for would be satisfied by making ignore_options also skip matching compiler option from special processing

Done in 2737d79e282de72d1b29128f437173e4892ced5e.

**Reply by 13steinj** [reply]
Wow; thank you for the prompt response and implementation!

> Ccache already has support for `-gsplit-dwarf` and caching the resulting file since [version 3.2.3](https://ccache.dev/releasenotes.html#_ccache_3_2_3).

Is there support for caching arbitrary byproducts rather than specifically `-gsplit-dwarf`? By which I mean, a byproduct perhaps generated by a compiler plugin?

**Reply by jrosdahl** [reply]
> Is there support for caching arbitrary byproducts rather than specifically `-gsplit-dwarf`? By which I mean, a byproduct perhaps generated by a compiler plugin?

There is also support for byproducts from `-ftest-coverage`, `-fstack-usage`, `--serialize-diagnostics` and `-Wa,-a=`, but there is no way to configure support for arbitrary byproducts.
